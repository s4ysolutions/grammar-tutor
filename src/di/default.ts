/*
 * Copyright 2022 by s4y.solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {DefaultPronounsDb} from '../tutor/pronouns/default';
import indexedDbFactory from '../kv/promise/indexedDB';
import {DefaultTutor} from '../tutor/tutor/default';
import {LearningDb, Tutor} from '../tutor';
import {KvPromiseLearningDb} from '../tutor/learned/kv-promise-db';
import {DefaultRouter} from '../router/default';
import {Router} from '../router';
import {DefaultUiState} from '../ui-state/default';
import {UiState} from '../ui-state';
import {Di} from './index';


class DefaultDi implements Di {
  private readonly _promiseKV = indexedDbFactory('sluchaj-zamenica');

  private readonly _pronounsDb = new DefaultPronounsDb();

  private readonly _learningDb = new KvPromiseLearningDb(this._promiseKV);

  private readonly _tutor = new DefaultTutor(this._pronounsDb, this._learningDb);

  private readonly _router = new DefaultRouter();

  private readonly _uiState = new DefaultUiState();

  get learningDb (): LearningDb {
    return this._learningDb;
  }

  get router (): Router {
    return this._router;
  }

  get tutor (): Tutor {
    return this._tutor;
  }

  get uiState (): UiState {
    return this._uiState;
  }

}

const singletonDi = new DefaultDi();

export const getDi = (): Di => singletonDi;
