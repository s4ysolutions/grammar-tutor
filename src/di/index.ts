/*
 * Copyright 2022 by s4y.solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {CasesInterrogativesDb, LearningProgress, NounsDb, Tutor} from '../tutor';
import {Router} from '../router';
import {UiState} from '../ui-state';

export interface Di {
  readonly personalPronounsDb: NounsDb
  readonly reflexivePronounsDb: NounsDb
  readonly possessivePronounsDb: NounsDb
  readonly interrogativePronounsDb: NounsDb
  readonly caseInterrogativesDb: CasesInterrogativesDb,
  readonly learningProgress: LearningProgress
  readonly tutor: Tutor
  readonly router: Router
  readonly uiState: UiState
}

export interface DiFactory {
  readonly di: Di;
}

let _di: Di = null;

export const initDi = (di: Di): void => {
  _di = di;
};

export const getDi = (): Di => {
  if (_di === null) {
    throw Error('Di is not initialized, call `initDi` firs');
  }
  return _di;
};

// default implem
