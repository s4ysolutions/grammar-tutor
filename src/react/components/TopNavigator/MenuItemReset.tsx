/*
 * Copyright 2022 by s4y.solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import log from '../../../log';
import T from '../../../l10n';
import {MenuItem} from '@mui/material';
import {getDi} from '../../../di/default';

const di = getDi();

const handleClick = (): void => {
  di.learningDb.reset().then();
  di.uiState.mainMenuOpen = false;
};

const MenuItemReset: React.FunctionComponent =
  (): React.ReactElement => {

    log.render('MenuItemReset');

    return <MenuItem onClick={handleClick}>
      {T`Reset`}
    </MenuItem>;
  };

export default MenuItemReset;
