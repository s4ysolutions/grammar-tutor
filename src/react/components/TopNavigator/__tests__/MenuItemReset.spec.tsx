/*
 * Copyright 2022 by s4y.solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import T from '../../../../l10n';
import memoryStoragePromiseFactory from '../../../../../tests/mocks/kv-promice/memoryStorage';
import MenuItemReset from '../menu-items/MenuItemReset';
import diFactory, {DefaultDi} from '../../../../di/default';
import {Di} from '../../../../di';
import {first} from 'rxjs';


describe('<MenuItemReset>', () => {

  beforeEach(() => {
    (diFactory as {di: Di}).di = new DefaultDi(memoryStoragePromiseFactory({}));
  });

  test('snapshot & observing', (done) => {
    const component = renderer.create(<MenuItemReset />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveProperty('props');

    diFactory.di.learningProgress.observableLessonStatistics()
      .pipe(first())
      .subscribe(statistics => {
        expect(statistics).toEqual({total: 0, wrong: 0});
        done();
      });
    renderer.act(() => {
      if ('props' in tree) {
        tree.props.onClick();
      }
    });

  });

  test('rendering & observing', (done) => {
    const { getByText } = render(<MenuItemReset />);
    const menuItem = getByText(T`Reset`);

    expect(menuItem).not.toBeNull();

    diFactory.di.learningProgress.observableLessonStatistics()
      .pipe(first())
      .subscribe(statistics => {
        expect(statistics).toEqual({total: 0, wrong: 0});
        done();
      });
    fireEvent.click(menuItem);
  });

});
