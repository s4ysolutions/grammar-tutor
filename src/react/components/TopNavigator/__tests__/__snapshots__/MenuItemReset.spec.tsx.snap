// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItemReset> render 1`] = `
<li
  className="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  Reset results
</li>
`;

exports[`<MenuItemReset> snapshot & observing 1`] = `
<li
  className="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  Reset results
</li>
`;

exports[`<MenuItemReset> snapshot 1`] = `
<li
  className="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  Reset results
</li>
`;
